### ISSUE TITLE
Reduce spike of network request on concurrent cache miss

### ISSUE THREAD

**Issue by alexvanin** [nspcc-dev/neofs-node#1248]
In https://github.com/nspcc-dev/neofs-node/pull/1236 we removed mutex from `lruNetCache` because it operates with thread-safe LRU cache. However spike of concurrent requests may produce many cache misses and trigger many network requests. 

We have a connection limit to NEO RPC node (https://github.com/nspcc-dev/neofs-node/pull/1031), so it should not break RPC node as it did before. Also, RPC node itself can cache the answer. However it seems sub optional and we can improve it with `RWMutex` tricks.

```go
func (c *Cache) get(key interface{}) (val interface{}, err err) {
    c.RLock() // use RLock() to avoid bottleneck of concurrent requests
    val, ok := c.cache.Peek(key)
    if ok && valid(val) {
        c.RUnlock()
        return val, nil
    }
    c.RUnlock()
   
    c.Lock()
    val, ok := c.cache.Peek(key) // check cache one more time in case of concurrent Lock()
    if ok && valid(val) {
        c.Unlock()
        return val, nil
    }

    val, err := c.GetFromNetwork(key)
    if err != nil {
        c.Unlock()
        return nil, err
    }
    c.cache.Add(key, val)
    c.Unlock()

    return val, nil
}   
```

In this issue try to benchmark both approaches with real neo-go RPC as well.

**Comment by fyrchik** [comment]
Here we want to:
1. Ensure that proper timeouts are set and used.
2. Ensure that TTL and size of caches can be modified in config.
3. Disallow concurrent network requests for the same container.
4. 
See #1583 for the stacktraces of hung goroutines.

**Comment by carpawell** [comment]
@alexvanin, @fyrchik, I tried the suggested approach in #2151, it also includes a test that allows switching test conditions.

Some common values for test variables and a new approach (container num was `105` in testnet):
<table>
  <tr>
    <th></th>
    <th>Num = 1</th>
    <th>Num = 10</th>
  </tr>
  <tr>
    <th>Same = false</th>
    <td>137.619185ms</td>
    <td>141.620603ms</td>
  </tr>
  <tr>
    <th>Same = true</th>
    <td>14.791988214s</td>
    <td>14.849283937s</td>
  </tr>
</table>

and an old one:

<table>
  <tr>
    <th></th>
    <th>Num = 1</th>
    <th>Num = 10</th>
  </tr>
  <tr>
    <th>Same = false</th>
    <td>444.614414ms</td>
    <td>446.490483ms</td>
  </tr>
  <tr>
    <th>Same = true</th>
    <td>397.617602ms</td>
    <td>401.071823ms</td>
  </tr>
</table>

It has both an advantage and a (critical) disadvantage. I also tried to implement some kind of optimization: block only the routines that want exactly a missing value (and do not block the others at all). It required code complication and some channel logic. Did not test it yet and am not sure if it deserves our efforts. What do you think?

**Comment by carpawell** [comment]
OK, i implemented the more complex solution in #2151, results are:

<table>
  <tr>
    <th></th>
    <th>Num = 1</th>
    <th>Num = 10</th>
  </tr>
  <tr>
    <th>Same = false</th>
    <td>247.84839ms</td>
    <td>254.156083ms</td>
  </tr>
  <tr>
    <th>Same = true</th>
    <td>83.1951ms</td>
    <td>86.990559ms</td>
  </tr>
</table>

I have changed my workplace and the internet prover since so i repeated the tests.
With the old approach (current `master` branch state):

<table>
  <tr>
    <th></th>
    <th>Num = 1</th>
    <th>Num = 10</th>
  </tr>
  <tr>
    <th>Same = false</th>
    <td>256.985444ms</td>
    <td>250.876516ms</td>
  </tr>
  <tr>
    <th>Same = true</th>
    <td>242.148676ms</td>
    <td>258.635695ms</td>
  </tr>
</table>

With the suggested in issue approach: 

<table>
  <tr>
    <th></th>
    <th>Num = 1</th>
    <th>Num = 10</th>
  </tr>
  <tr>
    <th>Same = false</th>
    <td>9.669001756s</td>
    <td>8.896387185s</td>
  </tr>
  <tr>
    <th>Same = true</th>
    <td>81.178283ms</td>
    <td>89.92683ms</td>
  </tr>
</table>

Ugly test placed [here](https://gist.github.com/carpawell/6e679b3eb3aee21472507a7ed48a4d69).
